numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -1481,6 +1481,26 @@ PyArray_EquivTypes(PyArray_Descr *type1, PyArray_Descr *type2)
 }
 
 
+/*
+ * This function returns true if a view can be safely created
+ * between the two types. This implies that PyArray_EquivTypes
+ * is true as well.
+ */
+NPY_NO_EXPORT unsigned char
+PyArray_ViewableTypes(PyArray_Descr *type1, PyArray_Descr *type2)
+{
+    if (!PyArray_EquivTypes(type1, type2)) {
+        return 0;
+    }
+    // a view of a stringdtype array has a corrupt arena, unless
+    // type1 and type2 are exactly the same object
+    if ((type1 != type2) && (type1->kind == 'T')) {
+        return 0;
+    }
+    return 1;
```

Comment:
Need to dig slightly deeper and use this pattern: https://github.com/numpy/numpy/blob/51fd714301f5fff8f8d6068951e0fdd44e98415d/numpy/_core/src/umath/ufunc_object.c#L790-L812

Making this not be string dtype specific!

Reply:
ah nice and then I can control how `view_offset` gets set in `string_to_string_resolve_descriptors`, that seems much more idiomatic

Reply:
So this idea doesn't quite work as cleanly as I'd hoped. If I set `view_offset != 0` in `string_to_string_resolve_descriptors`, there's an assert that triggers if `casting == NPY_NO_CASTING`:

https://github.com/numpy/numpy/blob/e1bf1d635016a0118f7d77bf8931071a2c74ad20/numpy/_core/src/multiarray/convert_datatype.c#L465-L482

In the latest version I'm about to push I guard that `assert` with an if statement that checks for stringdtype, but that seems unsatisfying. Maybe the correct thing is for stringdtype to set e.g. `NPY_EQUIV_CASTING` for casts with distinct allocators, but when I tried doing that it broke reductions and I couldn't see a clear path forward on fixing that, because the check for valid reductions relies on `PyArray_EquivTypes`, which enfoces `NPY_NO_CASTING` as the minimum casting level:

https://github.com/numpy/numpy/blob/e1bf1d635016a0118f7d77bf8931071a2c74ad20/numpy/_core/src/umath/ufunc_object.c#L2394-L2411

Maybe we need a new casting level between `NO_CASTING` and `EQUIV_CASTING` for this case where data can live outside the array buffer on the descriptor?

Reply:
I think we can probably just remove the assert, just need to make sure to not document anywhere that no-casting *implies* view (I doubt we use it anywhere, but not sure).  It used to, for these strings it can't since they use a whole new mechanism for storage... that seems fine.
(you could keep the assert together with checking that the dtype implements the hook for fixing the dtype in the array creation call, given that reason.)